client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java

```diff
@@ -148,60 +195,90 @@ protected AbstractRequestResponseClient(final Context context, final ProtonConne
+     */
+    protected final void createAndSendRequest(final String action, final JsonObject payload, final Handler<AsyncResult<R>> resultHandler) {
+        createAndSendRequest(action, null, payload, resultHandler);
     }
 
+    /**
+     * Creates a request message for a payload and headers and sends it to the peer.
+     * 
+     * @param action The operation that the request is supposed to trigger/invoke.
+     * @param properties The headers to include in the request message as AMQP application properties.
+     * @param payload The payload to include in the request message as a an AMQP Value section.
+     * @param resultHandler The handler to notify about the outcome of the request.
+     * @throws NullPointerException if any of action or result handler is {@code null}.
+     * @throws IllegalArgumentException if the properties contain any non-primitive typed values.
+     * @see AbstractHonoClient#setApplicationProperties(Message, Map)
+     */
     protected final void createAndSendRequest(final String action, final Map<String, Object> properties, final JsonObject payload,
                                       final Handler<AsyncResult<R>> resultHandler) {
 
+        Objects.requireNonNull(action);
+        Objects.requireNonNull(resultHandler);
+
         final Message request = createMessage(action, properties);
         if (payload != null) {
-            request.setContentType("application/json; charset=utf-8");
+            request.setContentType(RequestResponseApiConstants.CONTENT_TYPE_APPLICATION_JSON);
             request.setBody(new AmqpValue(payload.encode()));
         }
-        sendMessage(request, resultHandler);
+        sendRequest(request, resultHandler);
     }
 
     /**
-     * Send the Proton message to the endpoint link and call the resultHandler later with the result object.
-     *
-     * @param request The Proton message that was fully prepared for sending.
-     * @param resultHandler The result handler to be called with the response and the status of the request.
+     * Sends a request message via this client's sender link to the peer.
+     * <p>
+     * This method first checks if the sender has credits left and if not, immediately
+     * fails the result handler.
+     * 
+     * @param request The message to send.
+     * @param resultHandler The handler to notify about the outcome of the request.
      */
-    // TODO: improve so that the available credits are checked. Wait for enough credits before sending first.
-    protected final void sendMessage(final Message request, final Handler<AsyncResult<R>> resultHandler) {
+    private final void sendRequest(final Message request, final Handler<AsyncResult<R>> resultHandler) {
 
         context.runOnContext(req -> {
-            replyMap.put((String) request.getMessageId(), resultHandler);
-            sender.send(request);
+            if (sender.getCredit() > 0) {
+                replyMap.put((String) request.getMessageId(), resultHandler);
```

Comment:
Should there be a protection against sending a request multiple times?
Might be over-protected, since this method is private, though - but would be quite hard to find if this happens with extensions in the future...
WDYT?